include/DeterministicNoiseSimulator.hpp

```diff
@@ -0,0 +1,121 @@
+        Simulator(seed), qc(qc) {
+    }
+
+    DeterministicNoiseSimulator(std::unique_ptr<qc::QuantumComputation>& qc, const std::string& noise_effects, double noise_prob):
+        qc(qc) {
+        setNoiseEffects(noise_effects);
+        setAmplitudeDampingProbability(noise_prob);
+    }
+
+    void setAmplitudeDampingProbability(double cGateNoiseProbability) {
+        //The probability of amplitude damping (t1) often is double the probability , of phase flip, which is why I double it here
+        noise_probability                            = cGateNoiseProbability;
+        sqrt_amplitude_damping_probability           = {sqrt(noise_probability * 2), 0};
+        one_minus_sqrt_amplitude_damping_probability = {sqrt(1 - noise_probability * 2), 0};
+    }
+
+    std::map<std::string, std::size_t> Simulate([[maybe_unused]] unsigned int shots) override{};
+
+    std::map<std::string, double> DeterministicSimulate();
+
+    //    std::map<std::string, std::string> AdditionalStatistics() override {
+    //        return {
+    //                {"step_fidelity", std::to_string(step_fidelity)},
+    //                {"approximation_runs", std::to_string(approximation_runs)},
+    //                {"final_fidelity", std::to_string(final_fidelity)},
+    //        };
+    //    };
+
+    [[nodiscard]] std::string intToString(long target_number, char value) const;
+
+    void apply_det_noise_sequential(const qc::Targets& targets);
+
+    std::map<std::string, double> AnalyseState(dd::QubitCount nr_qubits, bool full_state) const;
+
+    void setNoiseEffects(const std::string& cGateNoise) { gate_noise_types = cGateNoise; }
+
+    [[nodiscard]] dd::QubitCount getNumberOfQubits() const override { return qc->getNqubits(); };
+
+    [[nodiscard]] std::size_t getNumberOfOps() const override { return qc->getNops(); };
+
+    [[nodiscard]] std::string getName() const override { return qc->getName(); };
+
+    const std::map<char, int> noise_effects = {
+            {'B', 2}, //Bit-flip
+            {'P', 2}, //Phase-flip
+            {'A', 2}, //Amplitude Damping
+            {'D', 4}, //Depolarisation
+    };
+
+    double           noise_probability = 0.0;
+    dd::ComplexValue sqrt_amplitude_damping_probability{};
+    dd::ComplexValue one_minus_sqrt_amplitude_damping_probability{};
+
+    dd::Package::mEdge density_root_edge;
+
+    bool weird_value_i_dont_understand_but_something_with_sequential = false;
+    char MeasureOneCollapsing(dd::Qubit index);
+
+private:
+    // TODO integrate the NoiseCacheTable into the ddPackage
+    static constexpr unsigned short NoiseSLOTS = 16384;
+    static constexpr unsigned short NoiseMASK  = NoiseSLOTS - 1;
+    struct NoiseEntry {
+        dd::Package::mNode *   a, *r; // a is the argument, r is the result
+        dd::ComplexValue       aw, rw;
+        dd::NoiseOperationKind which; // type of operation
+        signed char            usedQubits[std::numeric_limits<dd::Qubit>::max() + 1];
```

Comment:
Why is this a signed char array? shouldn't this probably be a boolean array, or even better a set of qubits, that is actually used?

Reply:
I need to specify to which qubits I am applying noise. In the worst-case, every qubit is being used, which is why I reserve space for each qubit. Usually, only 1 or 2 qubits are actually involved, however, at compile time I have to expect the worst case.